### DISCUSSION TITLE
HeaderedContentControl

### DISCUSSION THREAD

**Discussion by vladkryv** [AvaloniaUI/Avalonia#7875] [Q&A] [answered]
HeaderedContentControl is not displayed (Including content that is visible when replaced with to ContentControl), does it need to include styles in the App.axaml?

**Comment by maxkatz6** [comment, marked as answer]
> does it need to include styles in the App.axaml

Yes. Looks like it was overlooked before. This control is used as a base implementation for other controls, and probably wasn't intended to be used by itself.

So, you need to create your own style for this control and include in the App.axaml. Not sure if there can be any universal implementation of this style to be included in the Avalonia default themes.

**Reply by vladkryv** [reply]
It's a bit strange but ok, thanks for the answer.

**Reply by robloo** [reply]
I think a universal style could be included to Avalonia. It makes sense that this control is used by itself in some cases. Style could come from the default header styles in WinUI.

**Reply by dameng324** [reply]
Same issue lead me here, I also think a universal style could be included to Avalonia.

**Reply by timunie** [reply]
Contributions are welcome I think. I have my own style for this control, but I guess adoption the WinUI style would be fine.

**Reply by thevortexcloud** [reply]
I have thrown together a quick template for the Fluent theme here 

https://github.com/thevortexcloud/Avalonia/commit/75101c53f244ac170c35c8deeee20369645e99ab

If it it looks good I can clean it up and add the Simple theme too, and I am happy to contribute it.

**Reply by robloo** [reply]
Since UWP/WinUI don't have a HeaderedContentControl (Header property exists on most controls directly to flatten the visual tree) we can't copy the style from Fluent v1/v2. However, this was added to the Windows Community Toolkit. I would suggest using that as a base and comparing to what you have. The community toolkit version probably had several reviews and has been used a bit already.

https://github.com/CommunityToolkit/WindowsCommunityToolkit/blob/6bfe8a5d21feb8818a8048b9fc41a84a70af8b0f/Microsoft.Toolkit.Uwp.UI.Controls.Layout/HeaderedContentControl/HeaderedContentControl.xaml#L5

WPF also has this control -- though it seems too outdated now.

https://github.com/dotnet/wpf/blob/b9b48871d457fc1f78fa9526c0570dae8e34b488/src/Microsoft.DotNet.Wpf/src/Themes/XAML/HeaderedContentControl.xaml#L4

**Reply by thevortexcloud** [reply]
I have taken a quick look at that UWP template and it does not seem to be directly translatable. It looks like it auto rearranges itself based on the device orientation (which is why it has the grid layout). I can certainly replicate the layout with a grid but as far as I am aware Avalonia has no equivalent for `VisualStateManager` (psuedo properties can get kind of close though if a control supports it). There is also no orientation property on the control that could be used either. I am not aware of any control in the existing themes that does this either. But feel free to correct me on this.

The WPF template is much closer to what I did using a stack panel. Except I added a border so that could be styled.

**Reply by timunie** [reply]
in 11.0 we have OnFormFactor but this is probably too advanced.

**Reply by thevortexcloud** [reply]
I don't mind trying to implement some advanced stuff but it does seem somewhat overkill to just display some content with a header. That said from what I can tell `OnFormFactor` is just used to tell the difference between mobile and desktop platforms, not the orientation/aspect ratio of the display?

Theoretically we could also achieve similar functionality by adding an orientation property to the control, similar to the StackPanel. But that would still not auto switch based on the device orientation.

If it does not need the auto switching stuff I will probably leave it as it is and do the Simple theme. What I have already is fairly close to the WPF version (albeit with border controls).

**Reply by timunie** [reply]
I agree. Let's keep it simple. Prepare the PR imo. But please add HeaderTemplate in addition to ContentTemplate. 

To add: We also have classes which can be used. like: `HeaderedContentControl Classes="horizontal"` if needed, but can be done in a separate PR later.

**Reply by thevortexcloud** [reply]
Pull request added https://github.com/AvaloniaUI/Avalonia/pull/13028

**Reply by robloo** [reply]
Yea, I missed that it was automatically changing layout like that. Definitely seems overkill and actually not something most apps would want out of the box.

It might also be a good idea to have an orientation in the future though for people that want to stack vertically or horizontally (side by side). Probably a future feature request though.

**Reply by thevortexcloud** [reply]
https://github.com/AvaloniaUI/Avalonia/pull/13028 has now been merged. So this should no longer be an issue for anybody following this discussion.

**Reply by timunie** [reply]
thx for your contribution ❤️
